### ISSUE TITLE
"unused_mut" is not detected when variable type is mutable reference and the referent is mutated

### ISSUE THREAD

**Issue by QuestionDriven** [rust-lang/rust#88003]
#### Code:
```
fn main() {
        let mut point = &mut Point { x: 10, y: 20};
        point.x = 30; 
        println!("{:?}", point);
}

#[derive(Debug)]
struct Point {
    x: i32,
    y: i32,
}
```
([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f8fbe9c17a4a180fb1aed9e5054ca3a))

#### Current behavior:
no warning

#### Expected behavior:

```
warning: variable does not need to be mutable
 --> src/main.rs:2:13
  |
2 |         let mut point = &mut Point { x: 10, y: 20};
  |             ----^^^^^
  |             |
  |             help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default
```

**Comment by leonardo-m** [comment]
I think we should search systematically for all similar diagnostic holes/bugs regarding unnecessary muts. I filed two similar things for Clippy, but I think this kind of stuff is better in rustc.

**Comment by kpreid** [comment]
If this lint had existed, it would have helped me learn the rules around `mut <ident>` and `&mut` reborrowing faster.

**Comment by lukas-code** [comment]
This was fixed by https://github.com/rust-lang/rust/pull/110960 and the fix is included in Rust 1.71.0.
